test/unittest/tests/block/test_block_file_write.cpp

```diff
@@ -0,0 +1,199 @@
+    }
+
+    SECTION("Test os cache dirty max should be correctly be calling fsync")
+    {
+        // When this is set, when the file bytes written is greater an fsync should be called, and
+        // the file bytes written is set back to zero.
+        block->os_cache_dirty_max = 800;
+
+        // The first block write should succeed.
+        REQUIRE((__ut_block_write_off(session->get_wt_session_impl(), block, buf, &offset, &size,
+                  &checksum, false, false, false)) == 0);
+        validate_block_write(session->get_wt_session_impl(), block, offset, size, checksum,
+          expected_str, buf->size, expected_offset);
+        REQUIRE(block->fh->written == DEFAULT_BLOCK_SIZE);
+
+        // At this point the file written is greater than os_cache_dirty_max, make sure that
+        // the session flag must be set before the fh->written is cleared.
+        REQUIRE((__ut_block_write_off(session->get_wt_session_impl(), block, buf, &offset, &size,
+                  &checksum, false, false, false)) == 0);
+        validate_block_write(session->get_wt_session_impl(), block, offset, size, checksum,
+          expected_str, buf->size, expected_offset);
+        REQUIRE(block->fh->written == DEFAULT_BLOCK_SIZE * 2);
+
+        // Flag is now set, the block write should flushed with fsync.
+        F_SET(session->get_wt_session_impl(), WT_SESSION_CAN_WAIT);
+        REQUIRE((__ut_block_write_off(session->get_wt_session_impl(), block, buf, &offset, &size,
+                  &checksum, false, false, false)) == 0);
+        validate_block_write(session->get_wt_session_impl(), block, offset, size, checksum,
+          expected_str, buf->size, expected_offset);
+        REQUIRE(block->fh->written == 0);
+    }
+
+    SECTION("Test writes bigger than block")
+    {
+        // Perform a normal write.
+        REQUIRE((__ut_block_write_off(session->get_wt_session_impl(), block, buf, &offset, &size,
+                  &checksum, false, false, false)) == 0);
+        validate_block_write(session->get_wt_session_impl(), block, offset, size, checksum,
+          expected_str, buf->size, expected_offset);
+
+        /*
+         * Test what happens when a bigger buffer than the block allocation size. - Doesn't work is
+         * there a bug? std::string str2(DEFAULT_BLOCK_SIZE, 'a');
+         * REQUIRE(__wt_buf_initsize(session->get_wt_session_impl(), buf, DEFAULT_BLOCK_SIZE 2) ==
+         * 0); str2.copy(reinterpret_cast<char *>(const_cast<void *>(buf->data)), str2.length(), 0);
+         * std::cout << str2 << std::endl;
+         * REQUIRE((__ut_block_write_off(session->get_wt_session_impl(), block, buf, &offset, &size,
+         *   &checksum, false, false, false)) == 0);
+         * validate_block_write(session->get_wt_session_impl(), block, str2, offset, size, checksum, 2,
+         * buf);
+         */
+    }
+
+    /*
+     * Does this need a test? It is hard to test whether the function below respects this or not.
```

Comment:
The test with caller_locked=true is needed.  Actually it is easy to check whether caller_lock=true works. It is hard to test whether caller_lock=false works.
1. __wt_spin_lock(session, &block->live_lock);
2. __ut_block_write_off() with caller_locked = true.  If this does not hang __block_write_off() correctly did not lock the lock.
3. validate_block_write()
4. __wt_spin_trylock(session, &block->live_lock) should return EBUSY to show the lock is still locked.
5.  __wt_spin_unlock(session, &block->live_lock) to unlock the lock.

To test caller_locked=false you need two tests.
1. __wt_spin_trylock(session, &block->live_lock) returns 0 to show the lock was unlocked.
2. __wt_spin_unlock(session, &block->live_lock); to unlock it again.
3. __ut_block_off_write() with caller_locked = false.
4. validate_block_write()
5. __wt_spin_trylock(session, &block->live_lock) returns 0 to show the lock was unlocked.
6. __wt_spin_unlock(session, &block->live_lock) to unlock it again.
This shows the write worked but does not show the block was ever locked.

Something like the below but more complicated.
Thread 1.
1. __wt_spin_lock(session, &block->live_lock);
2. int expected_write_time = How long the write is expected to take.
3. pthread_create() thread2
4. start_time = read time.
5. __ut_block_write_off() with caller_locked=false
Thread 1 hangs here.
6.  end_time = read time.
7. WT_ASSERT(session, (end_time - start_time) > 2 * expected_write_time); to verify it hung.
8. validate_block_write()
9. __wt_spin_trylock(session, &block->live_lock) should return 0 to show the lock is unlocked.
10. __wt_spin_unlock(session, &block->live_lock);

Thread 2.
1. sleep(10 * expected_write_time);
2. __wt_spin_unlock(session, &block->live_lock);

I said something like the above since the above test would fail if right after pthread_create() there is a context switch from thread 1 to thread 2 and thread 2 completes both steps 1 and 2 before there is a context switch back to thread 1. In that case thread 1 would not hang and the difference end_time - start_time would be too small. This is unlikely in current multicore CPUs plus some OSs context switch during sleep. However it could happen. A more complicated test with critical sections could completely eliminate the false failure.